File: trafficcounter-backend/src/main/java/ch/bfh/trafficcounter/service/VehicleDataService.java
/*
 * Copyright 2023 Manuel Riesen, Sandro Rüfenacht, Sven Trachsel
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 */

package ch.bfh.trafficcounter.service;

import ch.bfh.trafficcounter.model.dto.HistoricDataCollectionDto;
import ch.bfh.trafficcounter.model.dto.geojson.GeoJsonFeatureCollectionDto;
import ch.bfh.trafficcounter.model.entity.MeasurementStatsType;

/**
 * Service for vehicle data and operations.
 *
 * @author Sven Trachsel
 */
public interface VehicleDataService {

    /**
     * Gets the amount of vehicle data as GeoJSON.
     *
     * @return vehicle data
     */
    GeoJsonFeatureCollectionDto getCurrentVehicleData();

    /**
     * Gets the amount of vehicles in the past from a specific measurement-point as GeoJSON
     *
     * @param measurementPointId the id of the selected measurement point
     * @param duration           duration from now backwards what data should be selected (24h or 7d)
     * @return vehicle data in the past duration
     * @throws IllegalArgumentException when an unsupported duration is entered
     */
    HistoricDataCollectionDto getHistoricalVehicleData(String measurementPointId, MeasurementStatsType duration) throws IllegalArgumentException;

    /**
     * Checks whether historic data for the measurement point is available
     *
     * @param id the id of the measurement point
     * @return true if data is available, false if not
     */
    boolean hasHistoricData(String id);

    /**
     * Initializes aggregated data for 24h and 7d
     * needed to load data after startup
     */
    void initializeAggregatedData();
}

File: trafficcounter-frontend/src/app/services/vehicle-data/vehicle-data.service.ts
/*
 * Copyright 2023 Manuel Riesen, Sandro Rüfenacht, Sven Trachsel
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE file or at
 * https://opensource.org/licenses/MIT.
 */

import {Injectable} from '@angular/core';
import {VehicleDataControllerService} from "../../api/services/vehicle-data-controller.service";
import {map, Observable} from "rxjs";
import {GeoJsonFeatureCollectionDto} from "../../api/models/geo-json-feature-collection-dto";
import {HistoricDataCollectionDto} from "../../api/models/historic-data-collection-dto";

@Injectable({
  providedIn: 'root'
})
export class VehicleDataService {

  constructor(private readonly vehicleDataControllerService: VehicleDataControllerService) {
  }

  /**
   * Gets the amount of vehicles from the API as GeoJSON.
   */
  public getVehicleData(): Observable<GeoJsonFeatureCollectionDto> {
    return this.vehicleDataControllerService.getCurrentVehicleData();
  }

  /**
   * Gets the amount of vehicles from the API as GeoJSON in a reactive way.
   */
  public getVehicleDataReactive(): Observable<GeoJsonFeatureCollectionDto> {
    return this.vehicleDataControllerService.getCurrentVehicleDataReactive();
  }

  /**
   * Gets the historical vehicle data from the API.
   * @param id measurement point ID
   * @param duration duration
   */
  public getHistoricalVehicleData(id: string, duration: string): Observable<HistoricDataCollectionDto> {
    return this.vehicleDataControllerService.getHistoricalVehicleData(id, duration)
      .pipe(map((historicalDataCollectionDto: HistoricDataCollectionDto) => {
        historicalDataCollectionDto.measurements.forEach(measurement => {
          if (typeof measurement.time === "string") {
            let timestampParts: number[] = measurement.time.split(/[- :]/).map(value => parseInt(value));
            measurement.time = new Date(timestampParts[0], timestampParts[1] - 1, timestampParts[2], timestampParts[3], timestampParts[4], timestampParts[5]);
          }
        });
        return historicalDataCollectionDto;
      }));
  }
}
